sdk/go/common/workspace/templates_zip.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workspace
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+)
+
+func isZIPTemplateURL(templateNamePathOrURL string) bool {
+	parsedURL, _ := url.Parse(templateNamePathOrURL)
+	return parsedURL.Path != "" && strings.HasSuffix(parsedURL.Path, ".zip")
+}
+
+func retrieveZIPTemplates(templateURL string) (TemplateRepository, error) {
+	var err error
+	// Create a temp dir.
+	var temp string
+	if temp, err = os.MkdirTemp("", "pulumi-template-"); err != nil {
+		return TemplateRepository{}, err
+	}
+
+	parsedURL, err := url.Parse(strings.ReplaceAll(templateURL, ".zip", ""))
+	if err != nil {
+		return TemplateRepository{}, err
+	}
+
+	var fullPath string
+	if fullPath, err = RetrieveZIPTemplateFolder(parsedURL, temp); err != nil {
+		return TemplateRepository{}, fmt.Errorf("failed to retrieve zip archive: %w", err)
+	}
+
+	return TemplateRepository{
+		Root:         temp,
+		SubDirectory: fullPath,
+		ShouldDelete: true,
+	}, nil
+}
+
+func RetrieveZIPTemplateFolder(templateURL *url.URL, tempDir string) (string, error) {
+	packageRequest, err := http.NewRequest(http.MethodPut, templateURL.String(), bytes.NewReader([]byte{}))
```

Comment:
MethodPut? I'd expect Get here?

Reply:
I've been using PUT for params in the API, but since we've refactored all of that state to path, I don't mind changing this.

Reply:
Though in our service's case, we do "upsert" - we create a new bundle each time.

Reply:
Feels more like an impl detail on our end, though - changing to GET